create documentation page describing the high-level picture of MOOSE backend libraries (#33152)
idaholab/moose: modules/doc/content/framework_development/computational_backends.md

> # Computational Backends and Accelerated Components

Reviewer: ```suggestion
# Computational Backends
```

libraries.

This page gives a high-level map of those paths and points to the detailed documentation for each
one.

## Framework view

The default MOOSE path uses MOOSE objects, libMesh data structures, and PETSc solvers. Optional
capabilities can replace or accelerate parts of that path:

```text
MOOSE application and input file
|
+-- Default MOOSE/libMesh path
|   +-- FE/FV variables, Kernels, BCs, Materials, UserObjects
|   +-- libMesh mesh and finite element data structures
|   +-- PETSc nonlinear and linear solvers
|
+-- Kokkos-MOOSE
|   +-- Kokkos implementations of selected MOOSE object systems
|   +-- GPU-portable assembly and execution for supported objects
|
+-- MFEM-MOOSE
|   +-- MFEMProblem, MFEMMesh, MFEM variables, kernels, BCs, and solvers
|   +-- MFEM finite element spaces and assembly levels
|   +-- Optional GPU and libCEED execution paths through MFEM
|
+-- NEML2
    +-- External material models connected through the [NEML2] block
    +-- Optional device-aware material evaluation
```

## Choosing a path

| Path | What it changes | Use it when | Main entry points |
| :- | :- | :- | :- |

> | Default MOOSE/libMesh | The standard MOOSE finite element and finite volume path | You need broad MOOSE capability, mature object coverage, or CPU-focused development | [Problem](syntax/Problem/index.md), [Kernels](syntax/Kernels/index.md), [Materials](syntax/Materials/index.md), [Executioner](syntax/Executioner/index.md) |

Reviewer: these main entry points are quite debatable. I would hope we don't have to dig into the Problem much to develop a simple app

let's give tutorials instead

| Kokkos-MOOSE | Implementations of selected MOOSE object systems for portable device execution | You want GPU-portable versions of supported MOOSE objects while keeping the MOOSE object model | [Kokkos-MOOSE](syntax/Kokkos/index.md), [Kokkos Kernels](syntax/KokkosKernels/index.md), [Kokkos Materials](syntax/KokkosMaterials/index.md) |
| MFEM-MOOSE | The finite element backend and the associated problem, mesh, variable, kernel, solver, and output objects | You need MFEM features such as high-order spaces, H(div)/H(curl) spaces, partial assembly, or MFEM GPU support | [MFEM-MOOSE](syntax/MFEM/index.md), [MFEMProblem](source/mfem/problem/MFEMProblem.md), [FESpaces](syntax/FESpaces/index.md) |
| NEML2 | Material model evaluation, not the finite element backend | You want MOOSE to call NEML2 material models and transfer inputs, outputs, and derivatives | [NEML2 syntax](syntax/NEML2/index.md) |

> | libCEED | Operator evaluation underneath MFEM, not a separate MOOSE object model | You are using MFEM-MOOSE and want libCEED-backed matrix-free execution where supported | [MFEM-MOOSE](syntax/MFEM/index.md) |

Reviewer: this seems like advanced use of MFEM-MOOSE>
only include it if we have a tutorial

Reviewer: I would think let's delete

## Default MOOSE/libMesh path

Most MOOSE applications use the default MOOSE/libMesh path. In this mode, application developers
derive from the standard MOOSE object families, such as `Kernel`, `ADKernel`, `IntegratedBC`,
`Material`, `UserObject`, `AuxKernel`, and related systems. MOOSE manages object construction,

> dependency ordering, residual and Jacobian assembly, output, restart, transfers, and execution.

Reviewer: MOOSE does all this for all the other backends too

> This path is the broadest and most mature route through the framework. It is usually the right

Reviewer: there is no route to the framework. framework is a toolkit to use to build your app

> starting point unless a problem specifically requires a newer accelerated path or a capability that is

Reviewer: define accelerated

problem, mesh, finite element space, variable, kernel, boundary condition, solver, executioner, and
output objects. This path is useful for capabilities that are natural in MFEM, including high-order
finite elements, H(div) and H(curl) spaces, partial assembly, and low-order-refined solvers.

MFEM-MOOSE is the path that most directly changes the finite element backend. It is also the MOOSE
path where libCEED-backed execution appears: when MFEM-MOOSE is built with libCEED support, MFEM
assembly and device options can use libCEED for supported matrix-free operations.

For details, see [Getting started with MFEM-MOOSE](syntax/MFEM/index.md).

## NEML2

NEML2 connects MOOSE simulations to NEML2 material models. It does not replace the MOOSE problem,
mesh, finite element, or solver stack. Instead, the `[NEML2]` block defines how MOOSE transfers
quantities into NEML2, retrieves outputs, and stores those outputs as MOOSE material properties.

Use NEML2 when the constitutive model itself should live in NEML2 or use NEML2 device scheduling. A
simulation can still otherwise follow the default MOOSE path or another compatible host path.

> For details, see [NEML2 syntax](syntax/NEML2/index.md).

Reviewer: I'd delete. It's not a backend, it's closer to an alternative materials system

## libCEED

libCEED is best understood as an acceleration layer used underneath supported finite element
operations, not as a user-facing MOOSE object family. In MOOSE, libCEED is currently exposed through
MFEM-MOOSE configuration. For example, MFEM executioners support assembly levels, and the `none`
assembly level is available when MFEM-MOOSE has been built with libCEED support.

Users generally should not start by asking whether to derive from libCEED classes. They should first
decide whether their problem belongs on the MFEM-MOOSE path, then select the MFEM assembly and device

> settings appropriate for their build and hardware.

Reviewer: I'd delete the whole thing. it s under MFEM, people using MFEM may pick various technology there

## Practical guidance

> Start with the default MOOSE/libMesh path when you need broad framework coverage or are building a

Reviewer: what do you mean by braod framework coverage?

why would you not use MFEM backend for a new application?

The advice on NEML2 is trivial

# Computational Backends

> MOOSE has several computational paths that share the same input-file-driven framework, but they do

Reviewer: I wouldnt mention the input file as the drive for the framework

computational paths seems weird. 
Maybe rephrase in an active way, where a modeler selects a backend for the right task

> not all replace the same layer of the stack. Some use the standard MOOSE object systems with libMesh

Reviewer: let's try to avoid introducing terminology we don't use much like describing moose as stacked layers

structures, some provide device-portable implementations of selected MOOSE object systems, and some
use a different finite element backend.

This page gives a high-level map of those choices and points to the detailed documentation for each
backend.

## Framework view

The default MOOSE path uses MOOSE objects, libMesh data structures, and PETSc solvers. Other
capabilities provide alternatives for parts of that path:

```text
MOOSE application and input file
|

> +-- Default MOOSE/libMesh path

Reviewer 2: I don't want to see the word default used. Default implies the system makes a choice. It doesn't. The input file specifications directly lead to the finite element backend used. I suppose that in the context of PETSc solvers, default is apt as PETSc does have default solvers when the user doesn't specify one.

|   +-- FE/FV variables, Kernels, BCs, Materials, UserObjects
|   +-- libMesh mesh and finite element data structures
|   +-- PETSc nonlinear and linear solvers
|
+-- Kokkos-MOOSE
|   +-- Kokkos implementations of selected MOOSE object systems

> |   +-- GPU-portable assembly and execution for supported objects

Reviewer: @NamjaeChoi or @lindsayad any preference on a bullet point here

Maybe to emphasize the compatibility? one can use the same type of mesh, and postprocess using the same auxkernels and PPs as regular moose

Reviewer 2: - Kokkos-MOOSE uses the libMesh finite element backend and consequently its `MeshBase` and `DofMap` data structures
- Many Kokkos-MOOSE system objects are compatible with host MOOSE/libMesh objects but there are exceptions. For instance host MOOSE strongly enforce nodal boundary conditions cannot be used in the same input as Kokkos-MOOSE residual objects

Reviewer: ```suggestion
|   +-- GPU-portable assembly and execution for supported objects
|   +-- Compatible with MOOSE/libMesh systems with a few exceptions
```

|
+-- MFEM-MOOSE
|   +-- MFEMProblem, MFEMMesh, MFEM variables, kernels, BCs, and solvers
|   +-- MFEM finite element spaces and assembly levels
|   +-- Optional GPU and libCEED execution paths through MFEM
```

## Choosing a path

| Path | What it changes | Use it when | Starting points |
| :- | :- | :- | :- |
| Default MOOSE/libMesh | Standard MOOSE finite element and finite volume object systems using libMesh data structures | You want to build with the object systems used by most MOOSE examples, modules, and applications | [Application development tutorial](getting_started/examples_and_tutorials/tutorial01_app_development/index.md), [framework examples](getting_started/examples_and_tutorials/examples_tutorials.md#framework-examples) |

> | Kokkos-MOOSE | Device-portable implementations of selected MOOSE object systems | You want supported MOOSE objects to execute through Kokkos while keeping the MOOSE object model | [Getting started with Kokkos-MOOSE](syntax/Kokkos/index.md), [Kokkos installation](getting_started/installation/install_kokkos.md) |

Reviewer: the justifications for moose and kokkos could be more compelling

| MFEM-MOOSE | MFEM-backed problem, mesh, variable, finite element space, kernel, solver, and output objects | You want to formulate the application using MFEM features such as high-order spaces, H(div)/H(curl) spaces, or partial assembly | [Getting started with MFEM-MOOSE](syntax/MFEM/index.md), [MFEM installation](getting_started/installation/install_mfem.md) |

## Default MOOSE/libMesh path

Most MOOSE applications use the default MOOSE/libMesh path. In this mode, application developers

> derive from the standard MOOSE object families, such as `Kernel`, `ADKernel`, `IntegratedBC`,

Reviewer: ```suggestion
create objects derived from the standard MOOSE object families, such as `Kernel`, `ADKernel`, `IntegratedBC`,
```

`Material`, `UserObject`, `AuxKernel`, and related systems.

> For details, see the

Reviewer: ```suggestion
For guidance on using the traditional MOOSE libMesh backend, see the
```

[application development tutorial](getting_started/examples_and_tutorials/tutorial01_app_development/index.md)
and the [framework examples](getting_started/examples_and_tutorials/examples_tutorials.md#framework-examples).

## Kokkos-MOOSE

Kokkos-MOOSE ports selected MOOSE object systems to GPU-portable execution while retaining the
general MOOSE programming model. A developer typically chooses Kokkos-MOOSE by using Kokkos-specific
object families, such as `Moose::Kokkos::Kernel`, `Moose::Kokkos::IntegratedBC`,
`Moose::Kokkos::Material`, or the corresponding Kokkos syntax pages.

Kokkos-MOOSE is not a separate finite element backend in the same sense as MFEM-MOOSE. It is a
device-portable implementation path for supported MOOSE systems, built around Kokkos data structures,

> memory-space rules, and parallel dispatch.

Reviewer 2: ```suggestion
memory-space rules, and parallel dispatch. It relies on libMesh as the finite element backend.
```

For details, see [Getting Started with Kokkos-MOOSE](syntax/Kokkos/index.md).

## MFEM-MOOSE

MFEM-MOOSE uses [MFEM](https://mfem.org) as its finite element backend. Inputs use MFEM-specific
problem, mesh, finite element space, variable, kernel, boundary condition, solver, executioner, and
output objects. This path is useful for capabilities that are natural in MFEM, including high-order
finite elements, H(div) and H(curl) spaces, partial assembly, and low-order-refined solvers.

> When MFEM-MOOSE is built with libCEED support, its `none` assembly level provides a matrix-free

Reviewer: none?

libCEED.

MFEM-MOOSE is the path that most directly changes the finite element backend.

> 

Reviewer 2: ```suggestion
```

# Computational Backends

A modeler can choose among several MOOSE computational backends depending on the problem being

> solved and the hardware being targeted. Some use the standard MOOSE object systems with libMesh data

Reviewer: standard probably same-ish connotation as default? @lindsayad where do you draw the line

Reviewer: traditional also in use in this page for moose-libmesh

Author: legacy?

Author: can we refer to the moose-libmesh duo as the legacy option?

Reviewer 2: I like traditional or historical. To me "legacy" means it's kind of deprecated and is no longer modern/first-class which is not the case

structures, some provide device-portable implementations of selected MOOSE object systems, and some
use a different finite element backend.

This page gives a high-level map of those choices and points to the detailed documentation for each
backend.

## Framework view

The MOOSE/libMesh path uses MOOSE objects, libMesh data structures, and PETSc solvers. Other
capabilities provide alternatives for parts of that path:

```text
MOOSE application and input file
|
+-- MOOSE/libMesh path
|   +-- FE/FV variables, Kernels, BCs, Materials, UserObjects
|   +-- libMesh mesh and finite element data structures
|   +-- PETSc nonlinear and linear solvers
|
+-- Kokkos-MOOSE
|   +-- Kokkos implementations of selected MOOSE object systems

> |   +-- libMesh finite element backend, MeshBase, and DofMap

Reviewer: ```suggestion
|   +-- libMesh finite element backend, MeshBase-derived mesh, and DofMap
```

+-- MFEM-MOOSE
|   +-- MFEMProblem, MFEMMesh, MFEM variables, kernels, BCs, and solvers
|   +-- MFEM finite element spaces and assembly levels

> |   +-- Optional GPU and libCEED execution paths through MFEM

Reviewer: I'd keep that line, as long as you can locate a case or a paper that did that

Reviewer: (the libCEED use)

Author: - test/tests/mfem/kernels/tests: MFEMDiffusionMatFree uses Executioner/device=ceed-cpu and Executioner/assembly_level=none.
  - test/tests/mfem/kernels/tests: MFEMHeatConductionMatFree does the same for transient heat conduction.
  - framework/doc/content/syntax/MFEM/index.md also documents that none assembly level is available with libCEED support.
